dataset/include/scipp/dataset/data_array.h

```diff
@@ -0,0 +1,109 @@
+// SPDX-License-Identifier: BSD-3-Clause
+// Copyright (c) 2021 Scipp contributors (https://github.com/scipp)
+/// @file
+/// @author Simon Heybrock
+#pragma once
+
+#include <string>
+
+#include "scipp/dataset/map_view.h"
+#include "scipp/variable/variable.h"
+
+namespace scipp::dataset {
+
+/// Policies for attribute propagation in operations with data arrays or
+/// dataset.
+enum class AttrPolicy { Keep, Drop };
+
+/// Data array, a variable with coordinates, masks, and attributes.
+class SCIPP_DATASET_EXPORT DataArray {
+public:
+  DataArray() = default;
+  DataArray(const DataArray &other, const AttrPolicy attrPolicy);
+  DataArray(const DataArray &other);
+  DataArray(DataArray &&other) = default;
+
+  DataArray(Variable data, Coords coords, Masks masks, Attrs attrs,
+            const std::string &name = "");
+  explicit DataArray(Variable data, typename Coords::holder_type coords = {},
+                     typename Masks::holder_type masks = {},
+                     typename Attrs::holder_type attrs = {},
+                     const std::string &name = "");
+
+  DataArray &operator=(const DataArray &other);
+  DataArray &operator=(DataArray &&other) = default;
+
+  explicit operator bool() const noexcept {
+    return m_data && m_data->operator bool();
+  }
+
+  const std::string &name() const;
+  void setName(const std::string &name);
+
+  const Coords &coords() const { return m_coords; }
+  // TODO either ensure Dict does not allow changing sizes, or return by value
+  // here
+  Coords &coords() { return m_coords; }
+
+  const Masks &masks() const { return *m_masks; }
+  Masks &masks() { return *m_masks; }
```

Comment:
Yes, this is a long-standing issue with most of `Variable` and `DataArray`. The main scenario where is used to arise was after a `std::move`. However, creating such variables or arrays from Python has never been possible, so in practice it was not much of an issue. Nevertheless it may be a good idea to consider changing this in all classes, but of course this is independent of this PR.